src/Compilers/CSharp/Portable/Symbols/Attributes/SourceAttributeData.cs

```diff
@@ -427,14 +455,31 @@ bool matches(byte[] targetSignature, ImmutableArray<ParameterSymbol> parameters,
             }
         }
 
+        internal override bool IsTargetAttribute(string namespaceName, string typeName)
+        {
+            return IsTargetAttribute(AttributeClass, namespaceName, typeName);
+        }
+
         /// <summary>
-        /// Gets the System.Type type symbol from targetSymbol's containing assembly.
+        /// Compares the namespace and type name with the attribute's namespace and type name.
+        /// Returns true if they are the same.
         /// </summary>
-        /// <param name="targetSymbol">Target symbol on which this attribute is applied.</param>
-        /// <returns>System.Type type symbol.</returns>
-        internal virtual TypeSymbol GetSystemType(Symbol targetSymbol)
+        internal static bool IsTargetAttribute(NamedTypeSymbol attributeClass, string namespaceName, string typeName)
         {
-            return targetSymbol.DeclaringCompilation.GetWellKnownType(WellKnownType.System_Type);
+            Debug.Assert(attributeClass is object);
+
+            if (!attributeClass.Name.Equals(typeName))
+            {
+                return false;
+            }
+
+            if (attributeClass.IsErrorType() && !(attributeClass is MissingMetadataTypeSymbol))
+            {
+                // Can't guarantee complete name information.
+                return false;
+            }
+
+            return attributeClass.HasNameQualifier(namespaceName);
```

Comment:
Should we also check the arity? I believe the compiler doesn't make use of generic attributes yet, so this logic may want to look for zero-arity attributes only.

Reply:
> Should we also check the arity? I believe the compiler doesn't make use of generic attributes yet, so this logic may want to look for zero-arity attributes only.

This is a good observation. However, this is not a new code, it was moved/copied from the base class. I would prefer to not make a change like that in context of this PR, especially that we would want it to be accompanied with new dedicated tests. Consider opening a separate issue for that. We probably should compare MetadataName  instead of name above. I am not sure about checking arity, we really do not care about that if name and signature match

Reply:
Opened #70554.